Design-Docs/standardizing-cluster-infrastructure.md

```diff
@@ -0,0 +1,277 @@
+
+- etcd encryption
+- HPA config
+- OIDC config
+- admission-control
+
+##### Layer 3: Kubernetes core
+
+There are a few components that have to be considered in this layer. First of all, the provider-specific cloud controller manager (CCM). The container network interface (CNI) is another component, which depends heavily on other selected components (Cilium, eBPF, node image, MTU size, configuration of Cluster API objects like `ProviderCluster`).
+
+There is also an interesting component, which is used for Kubernetes clusters that follow the zero-trust principle: the konnectivity-agent. It depends on the node image, as a konnectivity-server is needed, on the Kubeadm configuration of the Cluster API objects (egress config), and can depend on the CNI, as environments without kube-proxy are currently not supported.
+
+#### Example of a cluster stack
+
+This example is very high-level and cannot go in-depth. It is an example of a selection of components, which could serve as a base for a working cluster stack.
+
+**Pre-cluster state**: Node image with Ubuntu, cgroups V2, containerd, crun, eBPF support, node-problem-detector, and konnectivity-server.
+
+**Cluster state**: Kubeadm and cloud-init, private networks, etcd encryption, audit-logs, admission-control, HPA config, signed certificates and CSR controller, Kubernetes 1.23.6
+
+**Kubernetes core**: Cillium with kube-proxy, CCM, CSI, konnectivity-agent
+
+### Proposal 2: Cluster definition is independent of the provider
+
+We propose a templating logic for clusters, which aims for having only two parameters defining the cluster infrastructure: the names of the provider and of the cluster stack. Both parameters can be exchanged freely, as long as the provider implements a certain cluster stack. All the rest of the cluster definition remains unaffected.
+
+To achieve this, apart from defining cluster stacks as described in Proposal 1, the provider integrations of Cluster API have to be aligned. SCS should define certain parameters, e.g. `ProviderMachineTemplate`, and all provider integrations of SCS providers should have some kind of translation matrix so that this input can be converted into the relevant provider object, e.g. `HCloudMachineTemplate`.
+
+We want to leave out the details of how this translation from SCS-defined properties into the respective provider integrations works. It can be discussed in future meetings.
+
+### Proposal 3: Applications are configured independently of the cluster infrastructure
+
+The first step of achieving independence of the layers, the cluster stacks have to be aligned in a certain way, e.g. that a CSR controller is a must-have for running metrics-server securely. This has been described in Proposal 1.
+
+The second step consists of two tasks:
+
+- Defining an adapted SCS cloud provider interface that unifies the inputs for creating load balancers
+- Standardizing the naming of storage classes
```

Comment:
Storage classes might have the same name but could have a big difference in performance.

Should a default storage class not be enough and the rest is provider-specific?

At gridscale, we currently have 3 performance classes for storage.
A fourth one will be coming.
Others could have just one or 2 it 4.

Also, providers could offer something that others don't, like local NVME storage.

Reply:
That's really similar to the challenges of the [OpenStack flavor naming](https://github.com/SovereignCloudStack/Operational-Docs/blob/main/flavor-naming-draft.MD) and OpenStack volume type naming discussions (not sure if the latter has officially started yet).

Reply:
There are certainly different approaches here - @ajfriesen you pointed out one. We should probably also spend a close look at related discussions, as you say @horazont!